pyteal/ast/subroutine_test.py

```diff
@@ -1302,3 +1302,159 @@ def mySubroutine(a1, a2, a3, a4, a5, a6, a7, a8, a9, a10):
         actual, _ = declaration.__teal__(options)
         options.setSubroutine(None)
         assert actual == expected
+
+
+def test_docstring_parsing_with_different_format():
+    short_desc = "Example of a ABIReturnSubroutine with short description docstring."
+    a_doc = "an abi Uint64 value"
+    return_doc = "A PyTeal expression that sets output Uint64 value as argument a."
+    long_desc = """Example first line.
+
+    This is a second line.
+
+    This is a third line that's so long it has to wrap in order to fit properly
+    in a line of source code.
+    """
+    expected_long_desc = "Example first line.\nThis is a second line.\nThis is a third line that's so long it has to wrap in order to fit properly in a line of source code."
+
+    def documented_method(a: pt.abi.Uint64, *, output: pt.abi.Uint64):
+        return output.set(a)
+
+    # Google format
+    documented_method.__doc__ = f"""{short_desc}
+
+    Args:
+        a: {a_doc}
+
+    Returns:
+        {return_doc}
+    """
+
+    mspec_dict = pt.ABIReturnSubroutine(documented_method).method_spec().dictify()
+    assert mspec_dict["desc"] == short_desc
+    assert mspec_dict["args"][0]["desc"] == a_doc
+    assert mspec_dict["returns"]["desc"] == return_doc
+
+    # epy format
+    documented_method.__doc__ = f"""
+    {short_desc}
+
+    @param a: {a_doc} 
+    @return: {return_doc} 
+    """
+
+    mspec_dict = ABIReturnSubroutine(documented_method).method_spec().dictify()
+    assert mspec_dict["desc"] == short_desc
+    assert mspec_dict["args"][0]["desc"] == a_doc
+    assert mspec_dict["returns"]["desc"] == return_doc
+
+    # numpy format
+    documented_method.__doc__ = f"""{short_desc}
+
+    Parameters
+    ----------
+    a:
+        an abi Uint64 value
+    output:
+        {a_doc} 
+
+    Returns
+    -------
+    uint64
+        {return_doc}
+    """
+
+    mspec_dict = ABIReturnSubroutine(documented_method).method_spec().dictify()
+    assert mspec_dict["desc"] == short_desc
+    assert mspec_dict["args"][0]["desc"] == a_doc
+    assert mspec_dict["returns"]["desc"] == return_doc
+
+    # rst format
+    documented_method.__doc__ = f"""{short_desc}
+
+    :param a: {a_doc} 
+    :returns: {return_doc}
+    """
+
+    mspec_dict = ABIReturnSubroutine(documented_method).method_spec().dictify()
+    assert mspec_dict["desc"] == short_desc
+    assert mspec_dict["args"][0]["desc"] == a_doc
+    assert mspec_dict["returns"]["desc"] == return_doc
+
+    # Short and long descriptions
+    documented_method.__doc__ = f"""{long_desc}
+
+    :param a: {a_doc} 
+    :returns: {return_doc}
+    """
```

Comment:
is this really short and long description?

Reply:
doesnt look like it 🤔

Reply:
nvm, I think `long_desc` has multi-lines, and the first line can serve as short description, no worries.

Reply:
oh, i see its "both" by having no newline before it. See the comment below this one